st-ligatures-boxdraw-20200407-28ad288.diff

```diff
@@ -0,0 +1,276 @@
++		if (glyphs[i].mode & ATTR_WDUMMY)
++			continue;
++		if (glyphs[i].mode & ATTR_BOXDRAW) {
++			specidx++;
++			continue;
++		}
++
++		if (codepoints[i] != specs[specidx].glyph)
++			((Glyph *)glyphs)[i].mode |= ATTR_LIGA;
++
++		specs[specidx++].glyph = codepoints[i];
++	}
++
++	free(codepoints);
++}
++
++void hbtransformsegment(XftFont *xfont, const Glyph *string, hb_codepoint_t *codepoints, int start, int length) {
++	hb_font_t *font = hbfindfont(xfont);
++	if (font == NULL)
++		return;
++
++	Rune rune;
++	ushort mode = USHRT_MAX;
++	hb_buffer_t *buffer = hb_buffer_create();
++	hb_buffer_set_direction(buffer, HB_DIRECTION_LTR);
++
++	// Fill buffer with codepoints.
++	for (int i = start; i < (start+length); i++) {
++		rune = string[i].u;
++		mode = string[i].mode;
++		if (mode & ATTR_WDUMMY)
++			rune = 0x0020;
++		hb_buffer_add_codepoints(buffer, &rune, 1, 0, 1);
++	}
++
++	// Shape the segment.
++	hb_shape(font, buffer, NULL, 0);
++
++	// Get new glyph info.
++	hb_glyph_info_t *info = hb_buffer_get_glyph_infos(buffer, NULL);
++
++	// Write new codepoints.
++	for (int i = 0; i < length; i++) {
++		hb_codepoint_t gid = info[i].codepoint;
++		codepoints[start+i] = gid;
++	}
++
++	// Cleanup.
++	hb_buffer_destroy(buffer);
++}
+diff --git a/hb.h b/hb.h
+new file mode 100644
+index 0000000..a209238
+--- /dev/null
++++ b/hb.h
+@@ -0,0 +1,7 @@
++#include <X11/Xft/Xft.h>
++#include <hb.h>
++#include <hb-ft.h>
++
++void hbunloadfonts();
++void hbtransform(XftGlyphFontSpec *, const Glyph *, int, int, int);
++
+diff --git a/st.c b/st.c
+index 3e48410..073e4c7 100644
+--- a/st.c
++++ b/st.c
+@@ -2584,8 +2584,14 @@ draw(void)
+ 		cx--;
+ 
+ 	drawregion(0, 0, term.col, term.row);
++	// Draw current line to format ligatures properly.
++	xdrawline(term.line[term.c.y], 0, term.c.y, term.col);
++
+ 	xdrawcursor(cx, term.c.y, term.line[term.c.y][cx],
+ 			term.ocx, term.ocy, term.line[term.ocy][term.ocx]);
++	// If cursor was on a transformed glyph, we need to redraw the previous line
++	if (term.ocy != term.c.y && (term.line[term.ocy][term.ocx].mode & ATTR_LIGA))
++		xdrawline(term.line[term.ocy], 0, term.ocy, term.col);
+ 	term.ocx = cx, term.ocy = term.c.y;
+ 	xfinishdraw();
+ 	xximspot(term.ocx, term.ocy);
+diff --git a/st.h b/st.h
+index a1928ca..07ba678 100644
+--- a/st.h
++++ b/st.h
+@@ -11,7 +11,8 @@
+ #define DIVCEIL(n, d)		(((n) + ((d) - 1)) / (d))
+ #define DEFAULT(a, b)		(a) = (a) ? (a) : (b)
+ #define LIMIT(x, a, b)		(x) = (x) < (a) ? (a) : (x) > (b) ? (b) : (x)
+-#define ATTRCMP(a, b)		((a).mode != (b).mode || (a).fg != (b).fg || \
++#define ATTRCMP(a, b)		(((a).mode & (~ATTR_WRAP) & (~ATTR_LIGA)) != ((b).mode & (~ATTR_WRAP) & (~ATTR_LIGA)) || \
++				(a).fg != (b).fg || \
+ 				(a).bg != (b).bg)
+ #define TIMEDIFF(t1, t2)	((t1.tv_sec-t2.tv_sec)*1000 + \
+ 				(t1.tv_nsec-t2.tv_nsec)/1E6)
+@@ -33,6 +34,7 @@ enum glyph_attribute {
+ 	ATTR_WRAP       = 1 << 8,
+ 	ATTR_WIDE       = 1 << 9,
+ 	ATTR_WDUMMY     = 1 << 10,
+	ATTR_BOXDRAW    = 1 << 11,
```

Comment:
git patch flagged this as a corrupt patch but i fixed it by inserting one whitespace like this:

```suggestion
 	ATTR_BOXDRAW    = 1 << 11,
```